crates/namada/src/ledger/native_vp/masp.rs

```diff
@@ -304,110 +296,327 @@ where
+                .ok_or_else(|| {
+                    Error::NativeVpError(native_vp::Error::SimpleMessage(
+                        "Overflow in bundle balance",
+                    ))
+                })?;
+        }
+        // Maybe the asset type has no attached epoch
+        None if changed_balances.tokens.contains_key(&vin.asset_type) => {
+            let (token, denom, digit) =
+                &changed_balances.tokens[&vin.asset_type];
+            // Determine what the asset type would be if it were epoched
+            let epoched_asset_type =
+                encode_asset_type(token.clone(), *denom, *digit, Some(epoch))
+                    .wrap_err("unable to create asset type")?;
+            if conversion_state.assets.contains_key(&epoched_asset_type) {
+                // If such an epoched asset type is available in the
+                // conversion tree, then we must reject the unepoched
+                // variant
+                let error =
+                    Error::NativeVpError(native_vp::Error::SimpleMessage(
+                        "epoch is missing from asset type",
+                    ));
+                tracing::debug!("{error}");
+                return Err(error);
+            } else {
+                // Otherwise note the contribution to this transparent input
+                let amount =
+                    token::Amount::from_masp_denominated(vin.value, *digit);
+                *bal_ref = bal_ref
+                    .checked_sub(&ValueSum::from_pair(token.clone(), amount))
+                    .ok_or_else(|| {
+                        Error::NativeVpError(native_vp::Error::SimpleMessage(
+                            "Overflow in bundle balance",
+                        ))
+                    })?;
+            }
+        }
+        // unrecognized asset
+        _ => {
+            let error = Error::NativeVpError(native_vp::Error::SimpleMessage(
+                "Unable to decode asset type",
+            ));
+            tracing::debug!("{error}");
+            return Err(error);
+        }
+    };
+    Ok(())
+}
+
+// Handle transparent output
+fn validate_transparent_output(
+    out: &TxOut,
+    changed_balances: &mut ChangedBalances,
+    transparent_tx_pool: &mut I128Sum,
+    epoch: Epoch,
+    conversion_state: &ConversionState,
+) -> Result<()> {
+    // Non-masp destinations subtract from transparent tx pool
+    *transparent_tx_pool = transparent_tx_pool
+        .checked_sub(
+            &I128Sum::from_nonnegative(out.asset_type, out.value as i128)
+                .ok()
+                .ok_or_err_msg("invalid value or asset type for amount")?,
+        )
+        .ok_or_err_msg("Underflow in output subtraction")?;
+
+    let bal_ref = changed_balances
+        .post
+        .entry(out.address)
+        .or_insert(ValueSum::zero());
+
+    match conversion_state.assets.get(&out.asset_type) {
+        Some(((address, _asset_denom, digit), asset_epoch, _, _))
```

Comment:
same comment

Reply:
Resolved in https://github.com/anoma/namada/pull/3186#discussion_r1606996403 .